src/java.base/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -226,133 +195,125 @@ private MethodHandle specialize(MethodHandle leafHandle) {
 
         if (highLevelType.returnType() != void.class) {
             MethodHandle returnFilter = identity(highLevelType.returnType());
+            int retBufPos = -1;
+            long retBufReadOffset = -1;
             int retContextPos = 0;
             int retInsertPos = 1;
+            if (callingSequence.needsReturnBuffer()) {
+                retBufPos = 0;
+                retBufReadOffset = callingSequence.returnBufferSize();
+                retContextPos++;
+                retInsertPos++;
+                returnFilter = dropArguments(returnFilter, retBufPos, MemorySegment.class);
+            }
             returnFilter = dropArguments(returnFilter, retContextPos, Binding.Context.class);
             List<Binding> bindings = callingSequence.returnBindings();
             for (int j = bindings.size() - 1; j >= 0; j--) {
                 Binding binding = bindings.get(j);
-                returnFilter = binding.specialize(returnFilter, retInsertPos, retContextPos);
+                if (callingSequence.needsReturnBuffer() && binding.tag() == Binding.Tag.VM_LOAD) {
```

Comment:
Here follows a complex piece of code to make the return buffer for multi-registers return values work.

It is probably not worth looking at too much, since it will be replaced in a followup patch by direct ASM bytecode generation.